src/golang/cmd/executor/executor/dynamic_teardown.go

```diff
@@ -31,52 +33,62 @@ func (ex *DynamicTeardownExecutor) Run(ctx context.Context) error {
+		// We use go routines to delete the clusters in parallel.
+		go func(dynamicIntegration *models.Integration) {
+			log.Infof("Checking dynamic integration %s, whose terraform directory is %s", dynamicIntegration.Name, dynamicIntegration.Config[shared.K8sTerraformPathKey])
+			defer wg.Done() // decrement the WaitGroup counter when the goroutine completes
+
+			if err := dynamic.ResyncClusterState(ctx, dynamicIntegration, ex.IntegrationRepo, ex.Database); err != nil {
+				log.Error(errors.Wrap(err, "Failed to resync cluster state"))
+				return
+			}
 
-	if dynamicIntegration.Config[shared.K8sStatusKey] == string(shared.K8sClusterActiveStatus) {
-		lastUsedTimestampStr := dynamicIntegration.Config[shared.K8sLastUsedTimestampKey]
-		lastUsedTimestamp, err := strconv.ParseInt(lastUsedTimestampStr, 10, 64)
-		if err != nil {
-			return errors.Wrap(err, "Unable to cast last used timestamp to int64")
-		}
+			if dynamicIntegration.Config[shared.K8sStatusKey] == string(shared.K8sClusterActiveStatus) {
+				lastUsedTimestampStr := dynamicIntegration.Config[shared.K8sLastUsedTimestampKey]
+				lastUsedTimestamp, err := strconv.ParseInt(lastUsedTimestampStr, 10, 64)
+				if err != nil {
+					log.Error(errors.Wrap(err, "Unable to cast last used timestamp to int64"))
+					return
+				}
 
-		keepaliveStr := dynamicIntegration.Config["keepalive"]
-		keepalive, err := strconv.ParseInt(keepaliveStr, 10, 64)
-		if err != nil {
-			return errors.Wrap(err, "Unable to cast keepalive period to int64")
-		}
+				keepaliveStr := dynamicIntegration.Config["keepalive"]
+				keepalive, err := strconv.ParseInt(keepaliveStr, 10, 64)
+				if err != nil {
+					log.Error(errors.Wrap(err, "Unable to cast keepalive period to int64"))
+					return
+				}
 
-		currTimestamp := time.Now().Unix()
-		if (currTimestamp - lastUsedTimestamp) > keepalive {
-			log.Info("Reached keepalive threshold, tearing down the cluster...")
-			if err = dynamic.DeleteK8sCluster(
-				ctx,
-				// Perform pods status check because in case there are still pods running, we don't
-				// want them to be killed by the teardown cron job.
-				false,
-				&dynamicIntegration,
-				ex.IntegrationRepo,
-				ex.Database,
-			); err != nil {
-				return errors.Wrap(err, "Unable to delete dynamic k8s integration")
+				currTimestamp := time.Now().Unix()
+				if (currTimestamp - lastUsedTimestamp) > keepalive {
+					log.Info("Reached keepalive threshold, tearing down the cluster...")
+					if err = dynamic.DeleteK8sCluster(
+						ctx,
+						// Perform pods status check because in case there are still pods running, we don't
+						// want them to be killed by the teardown cron job.
+						false,
+						dynamicIntegration,
+						ex.IntegrationRepo,
+						ex.Database,
+					); err != nil {
+						log.Error(errors.Wrap(err, "Unable to delete dynamic k8s integration"))
+						return
+					}
+				} else {
+					log.Info("Have not reached keepalive threshold, not tearing down the cluster.")
+				}
```

Comment:
Now that it's a goroutine, should we have some timeout mechanism? Actually, was there a timeout mechanism before, or do our executors just run unbounded.

Reply:
Our executor runs unbounded, but in this case Terraform itself has timeout built in so we won't run forever even during the error case.

Reply:
got it. Can you note that at the beginning of the goroutine? It's one of the first things that comes to my mind when reading unbounded goroutines (usually I've been taught to use ErrGroups where all goroutines share the same context).